Remove IPVLAN support

pchaigno opened the pull request:
Following the deprecation of the IPVLAN support in v1.11, we can now remove the code.

```release-note
Remove IPVLAN support following the deprecation in v1.11.
```

pchaigno committed:
docs: Note on removal of IPVLAN in upgrade guide

Signed-off-by: Paul Chaignon <paul@cilium.io>

ldelossa reviewed (approved):
All that deleting much have felt great.

borkmann reviewed (approved):
Nice, thanks for removing 🚀

pchaigno committed:
loader: Allow TestCompileAndLoadHostEndpoint to work without IPVLAN

The TestCompileAndLoadHostEndpoint "unit" test is covering the loading
of the host datapath (reloadHostDatapath). Currently, when IPVLAN is
disabled, part of this code path consists in patching and loading a
second object file for the second host device (cilium_net). This part is
a trickier to test because it requires a bit more information. So the
test skips it by enabling IPVLAN.

A subsequent commit will remove all IPVLAN support, so we need to find
another way to make the test work. This commit achieves that by setting
up all configuration necessary for the patching and loading of
cilium_net's BPF programs to succeed.

The main struggle here is that instead of working from the object file
for cilium_host and patching it with the information for cilium_net, we
are working from an object file compiled with node_config.h's default
contents.

First, the node_config.h defines the endpoint ID as 65535 so we need
to assign that ID to the host endpoint to avoid errors such as:

    no such string \"test_cilium_calls_00001\" in ELF

Then, node_config.h adds a test_ prefix to map names so we need to
redefine map names in the agent to avoid the following errors. In order
to do that, we make callsmap.{Host,Netdev}MapName variables instead of
constants.

    failed to write ELF file: test_loader/bpf_host_cilium_net.o: no such string \"cilium_policy_65535\" in ELF
    failed to write ELF file: test_loader/bpf_host_cilium_net.o: no such string \"cilium_calls_hostns_65535\" in ELF

Of course, we also need to create the endpoint's state directory to be
able to write the patched object file there.

Finally, we need to make SECCTX_FROM_IPCACHE into a proper static data
variable in our node_config.h so that it appears as a symbol we can
replace in the object file.

With all that, we can patch and load the object file for cilium_net.

Signed-off-by: Paul Chaignon <paul@cilium.io>

tklauser commented on pkg/datapath/linux/config/config.go:
```diff
@@ -178,10 +177,6 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 	cDefinesMap["CT_TAIL_CALL_BUFFER6"] = "cilium_tail_call_buffer6"
 
-	if option.Config.DatapathMode == datapathOption.DatapathModeIpvlan {
-		cDefinesMap["ENABLE_EXTRA_HOST_DEV"] = "1"
```
It looks like this is the only place where `ENABLE_EXTRA_HOST_DEV` is defined, so this allows us to drop the `#else` part her in `bpf/bpf_host.c` as well:

https://github.com/cilium/cilium/blob/8f606bafb696076609a01f88204d895576043950/bpf/bpf_host.c#L446-L453

tklauser commented on api/v1/openapi.yaml:
```diff
@@ -2246,8 +2246,6 @@ definitions:
         type: integer
       datapathMode:
         "$ref": "#/definitions/DatapathMode"
-      ipvlanConfiguration:
```
I think we'll have to keep this field in the API, just no longer populate it. Otherwise we'de break backwards compatibility.

tklauser replied:
I just noticed that the field is marked as `omitempty` in the resulting Go code, so I think it's fine to remove it.

tklauser commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -1151,14 +1151,6 @@ ipv6:
   # -- Enable IPv6 support.
   enabled: false
 
-ipvlan:
```
Cannot comment on the actual line, but the documentation for the `datapathMode` helm value also needs an update:

https://github.com/cilium/cilium/blob/8f606bafb696076609a01f88204d895576043950/install/kubernetes/cilium/values.yaml.tmpl#L432-L435

Valid values will be `veth` and `lb-only` after this change, AFAIK.

pchaigno replied:
Yes, we've discussed that internally with Martynas and Daniel. This is broken already on master in several places today (as evidenced by the above comment). I'll do a follow up PR to fix that.

tklauser reviewed (changes requested):
> `44 files changed, 59 insertions(+), 1229 deletions(-)`

Nice change! Must feel good to remove that much code 😄

Some comments inline.

pchaigno committed:
bpf: Remove ENABLE_EXTRA_HOST_DEV

Because we removed the IPVLAN feature, ENABLE_EXTRA_HOST_DEV is now
never defined. We can thus remove all datapath code guarded by that
macro.

Suggested-by: Tobias Klauser <tobias@cilium.io>
Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
Remove IPVLAN code

IPVLAN support was deprecated in v1.11 and can now be fully removed.
This commit removes all IPVLAN code in the agent and in the datapath.
Additional simplifications of the code may be possible following this.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
bpf: Remove ENABLE_HOST_REDIRECT

Because we removed the IPVLAN feature, ENABLE_HOST_REDIRECT is now
always defined. We can thus remove all guards with that macro in the
datapath code.

Signed-off-by: Paul Chaignon <paul@cilium.io>

sayboras reviewed (approved):
✅
